NWC: Add final features

hydra-yse opened the pull request:
This PR:
- Adds support for the `make_invoice` command
- Adds an optional expiry per NWC connection
- Adds an optional periodic balance per NWC connection

## TO-DOs
- [x] Properly expose, without runtime errors, the `BreezNwcService` struct through FRB

roeierez reviewed (changes requested):
Looks good

dangeross commented on lib/plugins/nwc/src/utils.rs:
```diff
@@ -0,0 +1,8 @@
+use std::time::{SystemTime, UNIX_EPOCH};
```
web_time?

roeierez commented on lib/plugins/nwc/src/lib.rs:
```diff
@@ -225,9 +259,52 @@ impl SdkNwcService {
             }
         };
 
+        if client.connection.receive_only && !matches!(req.params, RequestParams::MakeInvoice(_)) {
```
Do we block also other commands such as ListTransactions on receive only?
Should we also return error to the caller here?

danielgranhao commented on lib/plugins/nwc/src/lib.rs:
```diff
@@ -330,11 +433,55 @@ impl SdkNwcService {
+            }
+            None => None,
+        }
+    }
```
We should test this carefully on Wasm node.js. I don't recall the details but according to #880 it may cause panics.

danielgranhao commented on lib/plugins/nwc/src/persist.rs:
```diff
@@ -26,30 +31,113 @@ impl Persister {
+        self.storage
+            .set_item(KEY_NWC_CONNECTIONS, serde_json::to_string(&connections)?)?;
+        Ok(())
+    }
```
Could we get races here that cause one of the connections to end up not persisted?

danielgranhao replied:
Same with `edit_nwc_connection` and `remove_nwc_connection`. Maybe this has been discussed in the past but I'm missing why it's safe.

hydra-yse commented on lib/plugins/nwc/src/persist.rs:
```diff
@@ -26,30 +43,142 @@ impl Persister {
+        }
+        self.storage.set_item(KEY_NWC_LOCK, 1.to_string())?;
+        Ok(Lock { p: self })
+    }
```
Yes, I was thinking as we did in the context of multiple SDK instances running (and therefore multiple NwcServices). In that case we cannot use a Mutex as it's not shared across plugin instances. But if that's overkill then yes, we can switch to a simple instance lock. Also yes, we should be careful about infinite locks, they can be set to 0 at every startup in case of a crash.

roeierez replied:
What are the consequences of this race? expiry/refresh is something we can get over by periodic query.

roeierez replied:
@dangeross  I discussed with @hydra-yse to change the implementation of set_item to optionally receive the old value as parameter and then only update if the old value equals to the current value ("get and set") which achieves atomicity without db locking.

hydra-yse commented:
@dangeross I've fixed the CI errors and tested on WASM, everything is working as expected. I've also added ConnectionExpired/Refreshed events for better UX

dangeross commented on lib/plugins/nwc/src/persist.rs:
```diff
@@ -26,30 +43,180 @@ impl Persister {
+
+    pub(crate) async fn get_min_interval(&self) -> Option<u64> {
+        let get_min_connection_interval = async || -> NwcResult<Option<u64>> {
+            let _ = self.lock().await?;
```
I think the race concern is then read/writing, so I don't think need to hold a lock when reading

dangeross commented on lib/plugins/nwc/src/persist.rs:
```diff
@@ -26,30 +43,180 @@ impl Persister {
+    }
+
+    pub(crate) async fn list_nwc_connections(&self) -> NwcResult<HashMap<String, NwcConnection>> {
+        let _ = self.lock().await?;
```
Same here, I'm not sure its needed. Only when we plan to add/edit/remove connections

hydra-yse commented on lib/core/src/plugin/storage.rs:
```diff
@@ -93,8 +96,28 @@ impl PluginStorage {
+                    return Err(PluginStorageError::DataTooOld);
+                }
+            }
+        }
```
Solved in 2b048b9

roeierez reviewed (approved):
One small issue to retry otherwise LGTM

dangeross reviewed (commented):
Looks good. ~Is the TODO in the PR still valid?~

roeierez commented on lib/core/src/plugin/storage.rs:
```diff
@@ -93,11 +104,35 @@ impl PluginStorage {
+        let mut con = persister.get_connection()?;
+        let tx = con.transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)?;
+        if let Some(old_value) = old_value {
+            if let Some(current_value) = Persister::get_cached_item_inner(&tx, &scoped_key)? {
```
I think the idea is that if we pass None we don't care about the old value, just simply update.

dangeross commented on lib/plugins/nwc/src/event.rs:
```diff
@@ -17,11 +17,18 @@ pub enum NwcEventDetails {
+    pub connection_name: Option<String>,
     pub event_id: Option<String>,
     pub details: NwcEventDetails,
 }
```
I see sometimes the connection name is set in NwcEvent, sometimes set in the NwcEventDetails. It feels a bit inconsistent. I think now you've added `connection_name` in NwcEvent it can be removed from ConnectionExpired and ConnectionRefreshed events (Alternatively only add it to the events it is set). Wdyt?

hydra-yse replied:
Makes sense, thanks Ross

erdemyerebasmaz commented on lib/plugins/nwc/src/model.rs:
```diff
@@ -0,0 +1,118 @@
+    /// If the expiry time is less than [crate::MIN_REFRESH_INTERVAL_SEC] seconds,
+    /// then it will take at most [crate::MIN_REFRESH_INTERVAL_SEC] seconds in order for the
+    /// connection to be deleted.
+    pub expiry_time_sec: Option<u32>,
```
Shouldn't we expose `expiration_timestamp` instead? How does this translate to the end-user? Do they have to calculate expiration time by adding up `created_at` & `expiry_time_sec`?

dangeross commented on lib/wasm/src/lib.rs:
```diff
@@ -403,4 +393,11 @@ impl BindingLiquidSdk {
     }
+
+    #[wasm_bindgen(js_name = "useNwcService")]
+    pub async fn use_nwc_service(&self, config: NwcConfig) -> WasmResult<BindingNwcService> {
```
```suggestion
    #[wasm_bindgen(js_name = "useNwcPlugin")]
    pub async fn use_nwc_plugin(&self, config: NwcConfig) -> WasmResult<BindingNwcPlugin> {
```

dangeross commented on packages/flutter_breez_liquid/rust/src/sdk.rs:
```diff
@@ -268,4 +260,10 @@ impl BreezSdkLiquid {
         self.sdk.disconnect().await
     }
+
+    pub async fn use_nwc_service(&self, config: NwcConfig) -> Result<BreezNwcService, SdkError> {
```
```suggestion
    pub async fn use_nwc_plugin(&self, config: NwcConfig) -> Result<BreezNwcPlugin, SdkError> {
```

dangeross commented on lib/core/src/sdk.rs:
```diff
@@ -5109,6 +5074,47 @@ impl LiquidSdk {
+        };
+        plugin.on_stop().await;
+        Ok(())
+    }
```
I don't see this being used

hydra-yse replied:
Yeah, it's just as another `stop` method to avoid calling `on_stop` from the plugin directly, or if a stop isn't exposed at all in the bindings. Also adds symmetry to the start_plugin call. You think it should be removed?

dangeross replied:
So it can be safely removed from the SDK? I just don't see it in the binding interfaces, or is it more something the plugin should be calling? (e.g. a disconnect() fn in the plugin should call it?)

dangeross reviewed (commented):
It looks very good. I'm mostly looking top level now, so consistent naming. I think it would be good to name the exposed NWC plugin and functions to access it more clearly. So `SdkNwcPlugin`, `BindingNwcPlugin`, `BreezNwcPlugin`, `use_nwc_plugin`. Just to really cement and make it clear how things fit together.
